[RecursiveASTVisitor] Assert we skip implicit instantiations.

hvdijk opened the pull request:
In DEF_TRAVERSE_TMPL_SPEC_DECL, we attempt to skip implicit instantiations by detecting that D->getTemplateArgsAsWritten() returns nullptr. Previously, this was not reliable. To ensure we do not regress, add an assertion and a test.

Sirraide commented on clang/include/clang/AST/RecursiveASTVisitor.h:
```diff
@@ -2069,22 +2069,24 @@ bool RecursiveASTVisitor<Derived>::TraverseTemplateArgumentLocsHelper(
-      TRY_TO(TraverseTemplateArgumentLocsHelper(                               \
-          ArgsWritten->getTemplateArgs(), ArgsWritten->NumTemplateArgs));      \
+       is the only callback that's made for this instantiation. */             \
+    if (TSK != TSK_ImplicitInstantiation) {                                    \
```
Do we have to check for `TSK_Undeclared` too? We do that in e.g. `TraverseFunctionHelper()`, which seems like a similar situation to this.

hvdijk replied:
Thanks. It doesn't seem like the same situation as `TraverseFunctionHelper` to me: there, the logic is that if it's an implicit instantiation, that instantiation isn't written in the source code anywhere, so should be skipped by AST traversal. But the check in `TraverseFunctionHelper` assumes we possibly did end up in an implicit instantiation anyway, and then it needs to be handled in some appropriate way. Still, I think you're probably right that this should check `TSK_Undeclared` as well. I don't think that can be covered by this clang-tidy check, but I can at least check that it doesn't break any other tests.

hvdijk replied:
Actually, we are in a loop going over the children of a `*TemplateSpecializationDecl`, is it possible to encounter `TSK_Undeclared` there in the first place? It's the process of instantiation that leads to a child of `*TemplateSpecializationDecl`, is it not? For `TSK_Undeclared`, I think that should not happen.

hvdijk replied:
But running with some extra debug checks clearly shows that that does happen, and therefore that my understanding is wrong. Okay then. However, no `TSK_Undeclared` has any `TemplateArgsAsWritten` in the testing that I ran, so all `TSK_Undeclared` were already implicitly skipped. I'm happy to explicitly skip them instead.

Sirraide replied:
I’m candidly not familiar w/ template instantiation to know if it *should* happen; to me it just seemed that the code in `TraverseFunctionHelper()` suggested that it might.

@erichkeane probably knows more about this.

hvdijk commented:
ping @sdkrystian 

FWIW, `getTemplateArgsAsWritten()` is implemented as

```c++
  /// Retrieve the template argument list as written in the sources,
  /// if any.
  const ASTTemplateArgumentListInfo *getTemplateArgsAsWritten() const {
    if (auto *Info = ExplicitInfo.dyn_cast<ExplicitInstantiationInfo *>())
      return Info->TemplateArgsAsWritten;
    return cast<const ASTTemplateArgumentListInfo *>(ExplicitInfo);
  }
```

and `ExplicitInfo` is defined as

```
  /// Further info for explicit template specialization/instantiation.
  /// Does not apply to implicit specializations.
  SpecializationOrInstantiationInfo ExplicitInfo = nullptr;
```

which does not say that this is guaranteed to be `nullptr` for implicit specializations, and suggests to me that for implicit instantiations, ignoring it like I do in this PR is the right thing to do. However, maybe instead it would be better so that implicit specializations never have any `ExplicitInfo`?

hvdijk committed:
[RecursiveASTVisitor] Assert we skip implicit instantiations.

In DEF_TRAVERSE_TMPL_SPEC_DECL, we attempt to skip implicit instantiations
by detecting that D->getTemplateArgsAsWritten() returns nullptr.
Previously, this was not reliable. To ensure we do not regress, add an
assertion and a test.

erichkeane reviewed (approved):
Please update the commit message/title in github to better reflect that you're ONLY adding the tests, else this LGTM.
